test/runnable/indexed_alias.d

```diff
@@ -0,0 +1,155 @@
+void test1()
+{
+    int[2] a = [1,2];
+    alias a1 = a[1];
+    assert(a1 == 2);
+    a1 = 42;
+    assert(a1 == 42);
+    assert(a1 == a[1]);
+    assert(a1 is a[1]);
+    assert(&a1 is &a[1]);
+}
+
+// set, get, identity, address to global static array elem
+void test2()
+{
+    static int[2] a = [1,2];
+    alias a1 = a[1];
+    assert(a1 == 2);
+    a1 = 42;
+    assert(a1 == 42);
+    assert(a1 == a[1]);
+    assert(a1 is a[1]);
+    assert(&a1 is &a[1]);
+}
+
+// set, get, identity, address to local dyn array elem
+void test3()
+{
+    int[] a = [1,2];
+    alias a1 = a[1];
+    assert(a1 == 2);
+    a1 = 42;
+    assert(a1 == 42);
+    assert(a1 == a[1]);
+    assert(a1 is a[1]);
+    assert(&a1 is &a[1]);
+}
+
+// set, get, from custom type w/ opover
+void test4()
+{
+    static struct Foo
+    {
+        string field = "42";
+        ref opIndex(T)(T)
+        {
+            return field;
+        }
+    }
+    Foo foo;
+    alias f1 = foo[1];
+    assert(f1 == "42");
+    f1 = "1337";
+    assert(f1 == "1337");
+    assert(foo.field == "1337");
+}
+
+// check for symbol conflict in different scopes
+int[2] c = [3,4];
+alias c0 = c[0];
+void test5()
+{
+    int[2] c = [101,202];
+    assert (c0 == 3);
+    c0 = 5;
+    assert (c[0] == 101);
+    assert (.c[0] == 5);
+}
+
+// indexed symbol is firstly indexed from a tuple
+void test6()
+{
+    alias Seq(T...) = T;
+    int[2] a = [0,1];
+    int[2] b = [2,3];
+    alias ab = Seq!(a,b);
+    alias a0 = ab[0][0];
+    alias a1 = ab[0][1];
+    alias b0 = ab[1][0];
+    alias b1 = ab[1][1];
+    assert(a0 == 0);
+    assert(a1 == 1);
+    assert(b0 == 2);
+    assert(b1 == 3);
+}
+
+// CTFE index
+void test7()
+{
+    static int index(){return 0;}
+    int[1] a = [123];
+    alias a0 = a[index()];
+    assert(a0 == 123);
+}
+
+// templatized alias
+void test8()
+{
+    int[2] a = [0,1];
+    alias accesElemOne(alias array) = array[0];
+    alias accesElemTwo(alias array) = array[1];
+    assert(accesElemOne!(a) == 0);
+    assert(accesElemTwo!(a) == 1);
+    alias acces(alias array, int index) = array[index];
+    assert(acces!(a,1) == 1);
+}
+
+// alias of the symbol created to tie the var to an index
+void test9()
+{
+    int[2] a = [0,1];
+    alias a1 = a[1];
+    alias aa1 = a1;
+    assert(aa1 == 1);
+}
+
+// use aliases instead of getters to extract RGBA components
+void test10()
+{
+    static struct Color
+    {
+        uint rgba;
+        ubyte opIndex(uint index)
+        in(index < 4)
+        {
+            return *((cast(ubyte*) &rgba) + index);
+        }
+    }
+    Color c = Color(0xDDCCEE00);
+    alias a = c[0];
+    alias r = c[1];
+    alias g = c[2];
+    alias b = c[3];
+    assert(a == 0);
+    assert(r == 0xEE);
+    assert(g == 0xCC);
+    assert(b == 0xDD);
+}
+
+void main()
+{
+    test1();
+    test2();
+    test3();
+    test4();
+    test5();
+    test6();
+    test7();
+    test8();
+    test9();
+    test10();
+}
```

Comment:
You'd need some tests with constant folding and CTFEs in the index.

Reply:
ctfe as index is verified in  runnable test7.